unison-sync/src/Unison/Sync/Types.hs

```diff
@@ -0,0 +1,506 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Unison.Sync.Types where
+
+import Data.Aeson
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import Data.Bifoldable
+import Data.Bifunctor
+import Data.Bitraversable
+import Data.ByteArray.Encoding (Base (Base64), convertFromBase, convertToBase)
+import Data.ByteString (ByteString)
+import Data.Map (Map)
+import Data.Map.NonEmpty (NEMap)
+import Data.Set (Set)
+import Data.Set.NonEmpty (NESet)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+
+-- | A newtype for JSON encoding binary data.
+newtype Base64Bytes = Base64Bytes ByteString
+
+instance ToJSON Base64Bytes where
+  toJSON (Base64Bytes bytes) = String . Text.decodeUtf8 $ convertToBase Base64 bytes
+
+instance FromJSON Base64Bytes where
+  parseJSON = Aeson.withText "Base64" $ \txt -> do
+    either fail (pure . Base64Bytes) $ convertFromBase Base64 (Text.encodeUtf8 txt)
+
+newtype RepoName = RepoName Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+newtype HashJWT = HashJWT Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+data HashJWTClaims = HashJWTClaims
+  { hash :: Base32,
+    entityType :: EntityType
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON HashJWTClaims where
+  toJSON (HashJWTClaims hash entityType) =
+    object
+      [ "h" .= hash,
+        "t" .= entityType
+      ]
+
+instance FromJSON HashJWTClaims where
+  parseJSON = Aeson.withObject "HashJWTClaims" $ \obj -> do
+    hash <- obj .: "h"
+    entityType <- obj .: "t"
+    pure HashJWTClaims {..}
+
+newtype Base32 = Base32 Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+newtype Hash = Hash Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON, ToJSONKey, FromJSONKey)
```

Comment:
When do we use which of these?

Reply:
* Show would just be for debugging
*  Eq is required by Ord
* Ord is required because Hashes are uses as a `NEMap` Key in the `DownloadEntitiesResponse`
* ToJSON and FromJSON are used to serialize/deserialize Hashes in TypedHashes and HashJWTClaims
* ToJSONKey and FromJSONKey are used to serialise/parse string-keys for the JSON object when serialising the `NEMap` in `DownloadEntitiesResponse`